pysam/libcalignmentfile.pyx

```diff
@@ -273,7 +307,193 @@ cdef bam_hdr_t * build_header_from_list(reference_names,
             raise MemoryError("could not allocate {} bytes".format(strlen(ctext), sizeof(char)))
         memcpy(dest.text, ctext, strlen(ctext))
 
-    return dest
+    return new_header
+
+
+cdef class AlignmentHeader(object):
+    """header information for a :class:`AlignmentFile` object"""
+
+    # See makeVariantHeader for C constructor
+    def __cinit__(self):
+        self.ptr = NULL
+
+    # Python constructor
+    def __init__(self):
+        self.ptr = bam_hdr_init()
+        if not self.ptr:
+            raise ValueError('cannot create AlignmentHeader')
+
+    def __dealloc__(self):
+        if self.ptr:
+            bam_hdr_destroy(self.ptr)
+            self.ptr = NULL
+
+    def __bool__(self):
+        return self.ptr != NULL
+
+    def copy(self):
+        return makeAlignmentHeader(bam_hdr_dup(self.ptr))
+
+    def clear(self):
+        """clear header.
+
+        Note that this will invalidate all objectst that reference the header.
+        """
+        if self.ptr:
+            bam_hdr_destroy(self.ptr)
+            self.ptr = NULL
+    
+    property nreferences:
+        """"int with the number of :term:`reference` sequences in the file.
+
+        This is a read-only attribute."""
+        def __get__(self):
+            if self.ptr:
+                return self.ptr.n_targets
+            else:
+                return 0
+
+    property references:
+        """tuple with the names of :term:`reference` sequences. This is a
+        read-only attribute"""
+        def __get__(self):
+            t = []
+            if self.ptr:
+                for x from 0 <= x < self.ptr.n_targets:
```

Comment:
probably also worth updating to `range()` syntax, rather than the archaic pyrex syntax.